### PR TITLE
Updating to the new architecture of FurnaceData

### PR DESCRIPTION
**Pull request by Feeliiix**
# The FurnaceData gets a full upgrade

## Why?
- The last code was very unstructured, for the same results many different ways was used.
- The values was picked very unthinkingly

#### - Now it has an architecture:
all values correspond to the underlaying values:

Underlaying Value | Value for Implementation
------------ | -------------
furnaceBurnTime | passedBurnTime (Thats not the same but it is calculated out of it)
currentItemBurnTime | maxBurnTime
cookTime | passedCookTime
totalCookTime | maxCookTime
#### - Now im checking the structure and the efficiency before any commit

### CODE REVIEW

File: src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java [hunk 1]
```diff
@@ -36,36 +35,37 @@
...
+     * @return The value for the already passed burn time
      */
-    MutableBoundedValue<Integer> remainingBurnTime();
+    MutableBoundedValue<Integer> passedBurnTime();
 
     /**
-     * Gets the {@link MutableBoundedValue} for the remaining cook time of the
-     * {@link Furnace}.
+     * Gets the {@link MutableBoundedValue} for the time the current fuel in the {@link Furnace} can
+     * burn totally.
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java]: This should probably say something like 'the maximum amount of time the furnace can burn for.'

File: src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java [hunk 2]
```diff
@@ -36,36 +35,37 @@
...
      *
-     * <p>The {@link #remainingCookTime()} is the difference of {@link #maxCookTime()} and
-     * the time the item has already been cooked.</p>
-     *
-     * @return The value for the remaining cook time
+     * @return The value for the time the current fuel can burn totally.
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java]: same here

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java [hunk 3]
```diff
@@ -28,46 +28,44 @@
 import org.spongepowered.api.data.manipulator.ImmutableDataManipulator;
 import org.spongepowered.api.data.manipulator.mutable.tileentity.FurnaceData;
 import org.spongepowered.api.data.value.immutable.ImmutableBoundedValue;
-import org.spongepowered.api.entity.Item;
 
 /**
- * An {@link ImmutableDataManipulator} representing the {@link Furnace}
- * information such as the {@link #remainingBurnTime()} and
- * {@link #remainingCookTime()}.
+ * An {@link ImmutableDataManipulator} representing the {@link Furnace} information such as the .
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java]: `such as the`...?

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java [hunk 4]
```diff
@@ -28,46 +28,44 @@
...
-     * Gets the {@link ImmutableBoundedValue} for the remaining cook time of the
-     * {@link Furnace}.
+     * Gets the {@link ImmutableBoundedValue} for the time the current fuel in the {@link Furnace}
+     * can burn totally.
      *
-     * <p>The {@link #remainingCookTime()} is the difference of {@link #maxCookTime()} and
-     * the time the item has already been cooked.</p>
-     *
-     * @return The immutable value for the remaining cook time
+     * @return The immutable value for the time the current fuel can burn totally.
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java]: No period for `@return` tags

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java [hunk 5]
```diff
@@ -28,46 +28,44 @@
...
     /**
-     * Gets the {@link ImmutableBoundedValue} for the cook time of the
-     * {@link Item} that should be cooked.
-     *
-     * <p>This is called "maxCookTime" because the client calculates
-     * {@link #remainingCookTime()} from the "maxCookTime" minus the
-     * time the item cooked already.</p>
+     * Gets the {@link ImmutableBoundedValue} for the already passed cook time of the {@link
+     * org.spongepowered.api.item.inventory.ItemStack} that cooks in the {@link Furnace}. When the
+     * this is equals the {@link #maxCookTime()} the {@link
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java]: `When the this is equals`, `When this is equal to` maybe?

File: src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java [hunk 6]
```diff
@@ -36,36 +35,43 @@
...
+     */
+    MutableBoundedValue<Integer> passedCookTime();
+
+    /**
+     * Gets the {@link MutableBoundedValue} for the total time the {@link
+     * org.spongepowered.api.item.inventory.ItemStack} have to cook until it is cooked.
      *
-     * <p>This is the maximum of {@link #remainingCookTime()}.</p>
+     * Its represented by the arrow icon in the {@link Furnace}, if the arrow is 100% filled the
+     * value is exact this one. So its the maximum of the {@link #passedCookTime()}.
```
Comment (Tzky) [inline src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java]: This value represents the progress of the arrow icon in the furnace. if the arrow is completely filled, it's value matches this value.

File: src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java [hunk 7]
```diff
@@ -36,36 +35,43 @@
...
      */
-    MutableBoundedValue<Integer> remainingCookTime();
+    MutableBoundedValue<Integer> maxBurnTime();
 
     /**
-     * Gets the {@link MutableBoundedValue} for the cook time of the
-     * {@link Item} that should be cooked.
+     * Gets the {@link MutableBoundedValue} for the already passed cook time of the {@link
+     * org.spongepowered.api.item.inventory.ItemStack} in the {@link Furnace}. When the {@link
+     * #passedCookTime()} is equals the {@link #maxCookTime()} the {@link
```
Comment (Tzky) [inline src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java]: When the passed cooktime equals the maxcooktime, the link...

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java [hunk 8]
```diff
@@ -28,46 +28,48 @@
...
- * {@link #remainingCookTime()}.
+ * An {@link ImmutableDataManipulator} representing the {@link Furnace}.
  */
 public interface ImmutableFurnaceData extends ImmutableDataManipulator<ImmutableFurnaceData, FurnaceData> {
 
     /**
-     * Gets the {@link ImmutableBoundedValue} for the remaining burn time of the
-     * {@link Furnace}.
+     * Gets the {@link ImmutableBoundedValue} for the already passed burn time of the {@link
+     * Furnace}. When the {@link #passedBurnTime()} is equals the {@link #maxBurnTime()}, the
```
Comment (Tzky) [inline src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java]: When the {@link #passedBurnTime()} is equal to the {@link #maxBurnTime()}

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java [hunk 9]
```diff
@@ -28,46 +28,48 @@
...
 
     /**
-     * Gets the {@link ImmutableBoundedValue} for the remaining cook time of the
-     * {@link Furnace}.
+     * Gets the {@link ImmutableBoundedValue} for the maximum amount of fuel that can be supplied
+     * with the used fuel item.
      *
-     * <p>The {@link #remainingCookTime()} is the difference of {@link #maxCookTime()} and
-     * the time the item has already been cooked.</p>
+     * Its represented by the flame icon in the {@link Furnace}, if the flame is 100% filled the
```
Comment (Tzky) [inline src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java]: Is represented by the flame icon .....

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java [hunk 10]
```diff
@@ -28,46 +28,48 @@
...
+     * fuel item
      */
-    ImmutableBoundedValue<Integer> remainingCookTime();
+    ImmutableBoundedValue<Integer> maxBurnTime();
 
     /**
-     * Gets the {@link ImmutableBoundedValue} for the cook time of the
-     * {@link Item} that should be cooked.
+     * Gets the {@link ImmutableBoundedValue} for the already passed cook time of the {@link
+     * org.spongepowered.api.item.inventory.ItemStack} in the {@link Furnace}. When this is equals
```
Comment (Tzky) [inline src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java]: When this is equal to

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java [hunk 11]
```diff
@@ -28,46 +28,48 @@
...
-     * time the item cooked already.</p>
-     *
-     * <p>This is the maximum of {@link #remainingCookTime()}.</p>
+     * @return The immutable value for the already passed cook time
+     */
+    ImmutableBoundedValue<Integer> passedCookTime();
+
+    /**
+     * Gets the {@link ImmutableBoundedValue} for the total time the {@link
+     * org.spongepowered.api.item.inventory.ItemStack} have to cook until it is cooked.
```
Comment (Tzky) [inline src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java]: the itemStack HAS to cook **or** the items have to cook.

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java [hunk 12]
```diff
@@ -28,46 +28,48 @@
...
      *
-     * @return The immutable value for the maximum cook time
+     * @return The immutable value for the time the item have to cook
```
Comment (Tzky) [inline src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java]: the time the item *has* to cook

File: src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java [hunk 13]
```diff
@@ -36,36 +35,42 @@
 public interface FurnaceData extends DataManipulator<FurnaceData, ImmutableFurnaceData> {
 
     /**
-     * Gets the {@link MutableBoundedValue} for the remaining burn time of the
-     * {@link Furnace}.
+     * Gets the {@link MutableBoundedValue} for the already passed burn time of the {@link Furnace}.
+     * When this is equals to the {@link #maxBurnTime()}, the current used fuel is depleted.
```
Comment (Tzky) [inline src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java]: When this is equal to

File: src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java [hunk 14]
```diff
@@ -36,36 +35,42 @@
...
+     * @return The value for the maximum amount of fuel that can be supplied with the used fuel item
      */
-    MutableBoundedValue<Integer> remainingCookTime();
+    MutableBoundedValue<Integer> maxBurnTime();
 
     /**
-     * Gets the {@link MutableBoundedValue} for the cook time of the
-     * {@link Item} that should be cooked.
+     * Gets the {@link MutableBoundedValue} for the already passed cook time of the {@link
+     * org.spongepowered.api.item.inventory.ItemStack} in the {@link Furnace}. When this is equals
```
Comment (Tzky) [inline src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java]: same here.

File: src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java [hunk 15]
```diff
@@ -36,36 +35,42 @@
...
-     * <p>This is called "maxCookTime" because the client calculates
-     * {@link #remainingCookTime()} from the "maxCookTime" minus the
-     * time the item cooked already.</p>
+     * @return The value for the already passed cook time
+     */
+    MutableBoundedValue<Integer> passedCookTime();
+
+    /**
+     * Gets the {@link MutableBoundedValue} for the total time the {@link
+     * org.spongepowered.api.item.inventory.ItemStack} have to cook until it is cooked.
```
Comment (Tzky) [inline src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java]: have / has

File: src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java (hunk 6 above)
Comment (Feeliiix) [inline src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java]: Thats nice, will update it @Tzky

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java (hunk 8 above)
Comment (Feeliiix) [inline src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java]: Think already changed.

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java (hunk 9 above)
Comment (Feeliiix) [inline src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java]: ?? You mena it's .... right?

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java (hunk 10 above)
Comment (Feeliiix) [inline src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java]: right

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java (hunk 11 above)
Comment (Feeliiix) [inline src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java]: Also right, my gramma skills :hankey:

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java (hunk 12 above)
Comment (Feeliiix) [inline src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java]: Also right...

File: src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java (hunk 13 above)
Comment (Feeliiix) [inline src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java]: right

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java (hunk 9 above)
Comment (Tzky) [inline src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java]: I was more like:
``This value represents...`` or ``This is representing...``

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java [hunk 16]
```diff
@@ -28,46 +28,48 @@
...
     /**
-     * Gets the {@link ImmutableBoundedValue} for the remaining cook time of the
-     * {@link Furnace}.
+     * Gets the {@link ImmutableBoundedValue} for the maximum amount of fuel that can be supplied
+     * with the used fuel item.
      *
-     * <p>The {@link #remainingCookTime()} is the difference of {@link #maxCookTime()} and
-     * the time the item has already been cooked.</p>
+     * It is represented by the flame icon in the {@link Furnace}, if the flame is 100% filled the
+     * value is exact this one. So its the maximum of the {@link #passedBurnTime()}.
```
Comment (Tzky) [inline src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java]: if the flame is 100% filled the value is exact**ly** this one, this makes it the maximum of the {@link #passedBurnTime()}.

Still a bit wonky (my try too).

File: src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java [hunk 17]
```diff
@@ -36,36 +35,42 @@
...
     /**
-     * Gets the {@link MutableBoundedValue} for the remaining cook time of the
-     * {@link Furnace}.
+     * Gets the {@link MutableBoundedValue} for the maximum amount of fuel that can be supplied with
+     * the used fuel item.
      *
-     * <p>The {@link #remainingCookTime()} is the difference of {@link #maxCookTime()} and
-     * the time the item has already been cooked.</p>
+     * It is represented by the flame icon in the {@link Furnace}, if the flame is 100% filled the
+     * value is exact this one. So its the maximum of the {@link #passedBurnTime()}.
```
Comment (Tzky) [inline src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/FurnaceData.java]: same as above.

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java (hunk 16 above)
Comment (Feeliiix) [inline src/main/java/org/spongepowered/api/data/manipulator/immutable/tileentity/ImmutableFurnaceData.java]: You're right

**Comment by Aaron1011** [comment]
Merged in https://github.com/SpongePowered/SpongeAPI/commit/0ce8b79bab0213a6a34c23896d51803e9b6dcd04.

**Pull request closed without merging**